src/psyclone/tests/psyir/transformations/inline_trans_test.py

```diff
@@ -1203,52 +1118,338 @@ def test_validate_named_arg(fortran_reader):
-    assert ("Failed to find the source for routine 'sub' and therefore "
-            "cannot inline it." in str(err.value))
+    assert call.routine.is_local
+    with pytest.raises(InternalError) as info:
+        _ = inline_trans._find_routine(call)
+    assert ("Failed to find the source code of the local routine 'sub'."
+            in str(info.value))
+
+
+def test_find_routine_unresolved_wildcard(fortran_reader):
+    '''Test that the routine can be found via a wildcard use statement.'''
+
+    wildcard_use = CALL_IN_SUB_USE.replace(", only : sub", "")
+    code = f"{wildcard_use}{SUB_IN_MODULE}"
+    psyir = fortran_reader.psyir_from_source(code)
+    call = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    assert call.routine.is_unresolved
+    result = inline_trans._find_routine(call)
+    assert isinstance(result, Routine)
+    assert result.name == "sub"
+
+
+def test_find_routine_unresolved(fortran_reader):
+    '''Test that the routine can be found when there are no use statements
+    and there is a 'raw' subroutine.
+
+    '''
+    code = f"{CALL_IN_SUB}{SUB}"
+    psyir = fortran_reader.psyir_from_source(code)
+    call = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    assert call.routine.is_unresolved
+    result = inline_trans._find_routine(call)
+    assert isinstance(result, Routine)
+    assert result.name == "sub"
+
+
+def test_find_routine_raw_to_module_exception(fortran_reader):
+    '''Test that the routine is not found and an exception is raised if
+    there is no use statement and the code for another routine with
+    the same name is specified within a module.
+
+    '''
+    code = f"{CALL_IN_SUB}{SUB_IN_MODULE}"
+    psyir = fortran_reader.psyir_from_source(code)
+    call = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    assert call.routine.is_unresolved
+    with pytest.raises(TransformationError) as info:
+        _ = inline_trans._find_routine(call)
+    assert ("Failed to find the source code of the unresolved routine "
+            "'inline_mod' after trying wildcard imports '[]' and all "
+            "routines that are not in containers." in str(info.value))
+
+
+def test_find_routine_unresolved_exception(fortran_reader):
+    '''Test that the expected exception is raised if the routine is
+    unresolved and can't be found in any wildcard imports or any 'raw'
+    subroutines.
+
+    '''
+    wildcard_use = CALL_IN_SUB_USE.replace(", only : sub", "")
+    code = (
+        f"{wildcard_use}"
+        f"module inline_mod\n"
+        f"end module inline_mod\n"
+        f"subroutine sub2()\n"
+        f"end subroutine sub2\n")
+    psyir = fortran_reader.psyir_from_source(code)
+    call = psyir.walk(Call)[0]
+    inline_trans = InlineTrans()
+    assert call.routine.is_unresolved
+    with pytest.raises(TransformationError) as info:
+        _ = inline_trans._find_routine(call)
+    assert ("Failed to find the source code of the unresolved routine 'sub2' "
```

Comment:
This has me stumped. AFAICS, `wildcard_use` only contains a call to 'sub' so why are we complaining that we can't find the source of 'sub2'?

    FileContainer[]
        Routine[name:'run_it']
            0: Call[name='sub']
                Reference[name:'a']
        Container[inline_mod]
        Routine[name:'sub2']

Reply:
Oops. There was an error in the exception string. Fixed now.